content/docs/08-troubleshooting/01-migration-guide/04-migration-guide-rsc-ui.mdx

```diff
@@ -0,0 +1,642 @@
+  const [_, setMessages] = useUIState();
+
+  return (
+    <div>
+      {flights.map(flight => (
+        <div
+          key={flight.id}
+          onClick={async () => {
+            const response = await sendMessage(
+              `I would like to choose flight ${flight.id}!`,
+            );
+
+            setMessages(msgs => [...msgs, response]);
+          }}
+        >
+          {flight.name}
+        </div>
+      ))}
+    </div>
+  );
+}
+```
+
+#### After: Use another chat hook with same ID from the component
+
+After switching to AI SDK UI, these messages are by initializing the `useChat` hook in the component with the same `id` as the parent component.
+
+```tsx filename="@/app/components/list-flights.tsx"
+'use client';
+
+import { useChat } from 'ai/react';
+
+export function ListFlights({ chatId, flights }) {
+  const { append } = useChat({
+    id: chatId,
+    body: { id: chatId },
+    maxSteps: 5,
+  });
+
+  return (
+    <div>
+      {flights.map(flight => (
+        <div
+          key={flight.id}
+          onClick={async () => {
+            await append({
+              role: 'user',
+              content: `I would like to choose flight ${flight.id}!`,
+            });
+          }}
+        >
+          {flight.name}
+        </div>
+      ))}
+    </div>
+  );
+}
+```
+
+### Saving Chats
+
+Before implementing `streamUI` as a server action, you should create an `<AI/>` provider and wrap your application at the root layout to sync between AI and UI states. During its initialization, you typically use the `onSetAIState` callback function to track updates of a mutable AI state and save it to the database when `done(...)` is called.
+
+#### Before: Save chats using callback function of context provider
+
+```ts filename="@/app/actions.ts"
+import { createAI } from 'ai/rsc';
+import { saveChat } from '@/utils/queries';
+
+export const AI = createAI({
+  initialAIState: {},
+  initialUIState: {},
+  actions: {
+    // server actions
+  },
+  onSetAIState: async ({ state, done }) => {
+    'use server';
+
+    if (done) {
+      await saveChat(state);
+    }
+  },
+});
+```
+
+#### After: Save chats using callback function of `streamText`
+
+When you switch over to AI SDK UI, you will save chats using the `onFinish` callback function of `streamText` present in the route handler.
+
+```ts filename="@/app/api/chat/route.ts"
+import { openai } from '@ai-sdk/openai';
+import { saveChat } from '@/utils/queries';
+import { streamText, convertToCoreMessages } from 'ai';
+
+export function POST(request) {
+  const { id, messages } = await request.json();
+
+  const coreMessages = convertToCoreMessages(messages);
+
+  const result = await streamText({
+    model: openai('gpt-4o'),
+    system: 'you are a friendly assistant!',
+    messages: coreMessages,
+    onFinish: async ({ responseMessages }) => {
+      try {
+        await saveChat({
+          id,
+          messages: [...coreMessages, ...responseMessages],
```

Comment:
what are `responseMessages?`

Reply:
https://sdk.vercel.ai/docs/ai-sdk-core/tools-and-tool-calling#response-messages